connect/proxy/listener_test.go

```diff
@@ -19,71 +19,99 @@ import (
 	"github.com/hashicorp/consul/lib/freeport"
 )
 
-func testSetupMetrics(t *testing.T) *metrics.InmemSink {
+type MockSink struct {
+	keys   [][]string
+	vals   []float32
+	labels [][]metrics.Label
+}
+
+func (m *MockSink) SetGaugeWithLabels(key []string, val float32, labels []metrics.Label) {
+	m.keys = append(m.keys, key)
+	m.vals = append(m.vals, val)
+	m.labels = append(m.labels, labels)
+}
+
+func (m *MockSink) IncrCounterWithLabels(key []string, val float32, labels []metrics.Label) {
+	m.keys = append(m.keys, key)
+	m.vals = append(m.vals, val)
+	m.labels = append(m.labels, labels)
+}
+
+// Empty methods below only included to satisfy MetricSink interface
+func (m *MockSink) EmitKey(key []string, val float32) {
+}
+func (m *MockSink) SetGauge(key []string, val float32) {
+}
+func (m *MockSink) IncrCounter(key []string, val float32) {
+}
+func (m *MockSink) AddSample(key []string, val float32) {
+}
+func (m *MockSink) AddSampleWithLabels(key []string, val float32, labels []metrics.Label) {
+}
+
+// Flattens the key for formatting along with its labels, removes spaces
+func (m *MockSink) flattenKeyLabels(parts []string, labels []metrics.Label) (string, string) {
+	buf := &bytes.Buffer{}
+	replacer := strings.NewReplacer(" ", "_")
+
+	if len(parts) > 0 {
+		replacer.WriteString(buf, parts[0])
+	}
+	for _, part := range parts[1:] {
+		replacer.WriteString(buf, ".")
+		replacer.WriteString(buf, part)
+	}
+
+	key := buf.String()
+
+	for _, label := range labels {
+		replacer.WriteString(buf, fmt.Sprintf(";%s=%s", label.Name, label.Value))
+	}
+
+	return buf.String(), key
+}
+
```

Comment:
If we are going to copy this code rather than attempt to make the one in `go-metrics` re-usable, It would be better put in some general test helper package so it can be used by other parts of Consul that need to make assertions about instrumentation.

We've historically not really done that but I think it's increasinly important that we don't break metrics as we refactor things so I think it would be useful. That would probably necessitate implementing all the methods too.

I'd still prefer actually that we just re-use the go-metrics one. I'll detail that in later comment. We shoudl be able to get that merged pretty quickly so maybe we should hold this PR til then (assuming it's days) if it ends up taking longer then maybe we can just add a comment here that this is temporary and land it as is?